test/utils/ktesting/signals.go

```diff
@@ -0,0 +1,48 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ktesting
+
+import (
+	"context"
+	"errors"
+	"os"
+	"os/signal"
+)
+
+var (
+	interruptCtx  context.Context
+	signalChannel chan os.Signal
+)
+
+func init() {
+	// Setting up signals is intentionally done in an init function because
+	// then importing ktesting in a unit or integration test is sufficient
+	// to activate the signal behavior.
+	signalCtx, _ := signal.NotifyContext(context.Background(), os.Interrupt)
+	cancelCtx, cancel := context.WithCancelCause(context.Background())
+	go func() {
+		<-signalCtx.Done()
+		cancel(errors.New("received interrupt signal"))
+	}()
+	interruptCtx = cancelCtx
+
+	signalChannel = make(chan os.Signal, 1)
+	// progressSignals will be empty on Windows.
+	if len(progressSignals) > 0 {
+		signal.Notify(signalChannel, progressSignals...)
```

Comment:
This is incomplete: receiving such a signal currently doesn't do anything. The intention is to implement the Gingko progress notification mechanism such that Gomega picks it up.

Progress reports have to go to an io stream where they are visible in plain `go test` without `-v`.

Reply:
I've implemented this. It's a bit less useful (but also simpler) than the code in Ginkgo. "go test -v" has to be used to see the output because "go test" redirects `io.Stderr`:

```console
$ go test -v -tags=example ./test/utils/ktesting/examples/gomega
=== RUN   TestGomega
You requested a progress report.

Expected
    <int>: 42
to equal
    <int>: 1
^C    example_test.go:43: ERROR: 
        Context was cancelled after 39.738s.
        Expected
            <int>: 42
        to equal
            <int>: 1
--- FAIL: TestGomega (39.74s)
```

The progress report was triggered with `killall -USR1 gomega.test`.